src/mlpack/methods/reinforcement_learning/environment/lunarlander.hpp

```diff
@@ -0,0 +1,232 @@
+/**
+ * @file lunarlander.hpp
+ * @author Xiaohong Ji
+ *
+ * This file is an implementation of Acrobot task:
+ * https://github.com/zippkidd/lunarLander
+ * https://gym.openai.com/envs/LunarLanderContinuous-v2/
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_ENVIRONMENT_LUNARLANDER_HPP
+#define MLPACK_METHODS_RL_ENVIRONMENT_LUNARLANDER_HPP
+
+#include <mlpack/core.hpp>
+
+namespace mlpack{
+namespace rl{
+
+/**
+ * Implementation of Lunarlander game. Lunarlander is a 2-link pendulum with only the
+ * second joint actuated. Intitially, both links point downwards. The goal is
+ * to swing the end-effector at a height at least the length of one link above
+ * the base. Both links can swing freely and can pass by each other, i.e.,
+ * they don't collide when they have the same angle.
+ */
+class LunarLander
```

Comment:
Let me know once you think the lunar lander code is ready or if you need any help from my side.

Reply:
Thank you, I am writing the OpenAI environment setting. The implementation of lunarlander refer to large box2d codes, so I am reading the code the figure out how to implement in C++.

Reply:
This implementation was referring to [this](https://github.com/zippkidd/lunarLander).  I don't it is good to use this implementation because we don't know the detailed definition of reward and etc. To make consistent with current mainstream setting, I am implementing the OpenAI settings.

Reply:
I don't mind to implement a variation of the lunar lander env, as we only like to have another  continuous env for testing.

Reply:
Okay, I got it. So, what about simplifying the gym setting, and refer to the [zippkidd](https://github.com/zippkidd/lunarLander) implementation. The implementation will look like that we only control the Lunarlander in the vertical direction. The action will be the amount of fuel to use, we use it to control the speed. When the height is zero, we begin to call the touchdown function to determine the final speed. Then we use the final speed to indicate how well the system landing is.  I will use speed as an indicator to assign reward according to the [piecewise function](https://github.com/zippkidd/lunarLander/blob/master/LanderSim.cpp#L116)

Does this make sense? Can you have a look whenever you have time? 

One more thing, for the simulation discussed in IRC, do you mean the we simulate it by human hands?

Reply:
> Does this make sense? Can you have a look whenever you have time?

Sounds good to me, let's see if PPO is able to solve the env.

> One more thing, for the simulation discussed in IRC, do you mean the we simulate it by human hands?

The keyboard agent? I just referenced it, to get familiar with the env, we don't use it for training.

Reply:
Oh, I mean how to set threashold when we define the PPO works like other env? Most of the time, I refer to this [web pages](https://github.com/openai/gym/wiki/Leaderboard). I do not have a good way to get the threashold.

Reply:
That depends on the initial start reward, the reward you add if e.g. ``velocity == 0``, do you give a negative reward per action as well? An example would be: initial reward 400, -0.5 per action:

``velocity > 0.0 && velocity < 2.0`` -> reward +100
``velocity >= 2.0 && velocity < 5.0`` -> reward +50
``velocity >= 5.0 && velocity < 10.0`` -> reward -10
``velocity >= 10.0 && velocity < 30.0`` -> reward -30

maybe we can come up with a reward that depends on the velocity, something like reward = -velocity could work, but ideally we add a positive reward (e.g. 100) if ``velocity > 0.0 && velocity < 2.0``.

main engine -> reward -0.3

so I think a reasonable threshold would be 400, the reward we get for a perfect landing should account for the actions (left/right, main engine). We could start with that and train PPO with it and see if the threshold looks reasonable. Let me know what you think.